pkg/logql/syntax/visit.go

```diff
@@ -0,0 +1,285 @@
+	VisitVectorAggregationFn      func(v *DepthFirstTraversal, e *VectorAggregationExpr)
+}
+
+// VisitBinOp implements RootVisitor.
+func (v *DepthFirstTraversal) VisitBinOp(e *BinOpExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitBinOpFn != nil {
+		v.VisitBinOpFn(v, e)
+	} else {
+		e.SampleExpr.Accept(v)
+		e.RHS.Accept(v)
+	}
+}
+
+// VisitDecolorize implements RootVisitor.
+func (v *DepthFirstTraversal) VisitDecolorize(e *DecolorizeExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitDecolorizeFn != nil {
+		v.VisitDecolorizeFn(v, e)
+	}
+}
+
+// VisitDropLabels implements RootVisitor.
+func (v *DepthFirstTraversal) VisitDropLabels(e *DropLabelsExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitDecolorizeFn != nil {
+		v.VisitDropLabelsFn(v, e)
+	}
+}
+
+// VisitJSONExpressionParser implements RootVisitor.
+func (v *DepthFirstTraversal) VisitJSONExpressionParser(e *JSONExpressionParser) {
+	if e == nil {
+		return
+	}
+	if v.VisitJSONExpressionParserFn != nil {
+		v.VisitJSONExpressionParserFn(v, e)
+	}
+}
+
+// VisitKeepLabel implements RootVisitor.
+func (v *DepthFirstTraversal) VisitKeepLabel(e *KeepLabelsExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitKeepLabelFn != nil {
+		v.VisitKeepLabelFn(v, e)
+	}
+}
+
+// VisitLabelFilter implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLabelFilter(e *LabelFilterExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLabelFilterFn != nil {
+		v.VisitLabelFilterFn(v, e)
+	}
+}
+
+// VisitLabelFmt implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLabelFmt(e *LabelFmtExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLabelFmtFn != nil {
+		v.VisitLabelFmtFn(v, e)
+	}
+}
+
+// VisitLabelParser implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLabelParser(e *LabelParserExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLabelParserFn != nil {
+		v.VisitLabelParserFn(v, e)
+	}
+}
+
+// VisitLabelReplace implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLabelReplace(e *LabelReplaceExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLabelReplaceFn != nil {
+		v.VisitLabelReplaceFn(v, e)
+	}
+}
+
+// VisitLineFilter implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLineFilter(e *LineFilterExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLineFilterFn != nil {
+		v.VisitLineFilterFn(v, e)
+	} else {
+		e.Left.Accept(v)
+		e.Or.Accept(v)
+	}
+}
+
+// VisitLineFmt implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLineFmt(e *LineFmtExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLineFmtFn != nil {
+		v.VisitLineFmtFn(v, e)
+	}
+}
+
+// VisitLiteral implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLiteral(e *LiteralExpr) {
+	if e == nil {
+		return
+	}
+	if v.VisitLiteralFn != nil {
+		v.VisitLiteralFn(v, e)
+	}
+}
+
+// VisitLogRange implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLogRange(e *LogRange) {
+	if e == nil {
+		return
+	}
+	if v.VisitLogRangeFn != nil {
+		v.VisitLogRangeFn(v, e)
+	} else {
+		e.Left.Accept(v)
```

Comment:
Only left? Is the another `LogRange` child?

Reply:
Yes, there is only a Left, no other children.